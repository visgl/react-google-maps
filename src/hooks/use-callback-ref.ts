import {Ref, useCallback, useState} from 'react';

export function useCallbackRef<T>() {
  const [el, setEl] = useState<T | null>(null);
  const ref = useCallback((value: T) => setEl(value), [setEl]);

  return [el, ref as Ref<T>] as const;
}
