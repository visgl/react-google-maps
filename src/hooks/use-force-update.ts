import {useReducer} from 'react';

export function useForceUpdate(): () => void {
  const [, forceUpdate] = useReducer(x => x + 1, 0);

  return forceUpdate;
}
