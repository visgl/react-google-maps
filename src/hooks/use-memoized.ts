import {useRef} from 'react';

export function useMemoized<T>(value: T, isEqual: (a: T, b: T) => boolean): T {
  const ref = useRef<T>(value);

  if (!isEqual(value, ref.current)) {
    ref.current = value;
  }

  return ref.current;
}
