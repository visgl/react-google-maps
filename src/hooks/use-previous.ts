import {useEffect, useRef} from 'react';

/**
 * A hook to store the previous value of a variable.
 * @param value The value to store
 * @returns The previous value
 */
export function usePrevious<T>(value: T): T | undefined {
  const ref = useRef<T>(undefined);
  useEffect(() => {
    ref.current = value;
  });
  // eslint-disable-next-line react-hooks/refs
  return ref.current;
}
