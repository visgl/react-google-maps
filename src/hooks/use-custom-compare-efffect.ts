import {DependencyList, EffectCallback, useEffect} from 'react';
import {useMemoized} from './use-memoized';

export function useCustomCompareEffect<T extends DependencyList>(
  effect: EffectCallback,
  dependencies: T,
  isEqual: (a: T, b: T) => boolean
) {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(effect, [useMemoized(dependencies, isEqual)]);
}
