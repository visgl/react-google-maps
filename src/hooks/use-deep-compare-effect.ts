import {DependencyList, EffectCallback} from 'react';
import {useCustomCompareEffect} from './use-custom-compare-efffect';
import isDeepEqual from 'fast-deep-equal';

export function useDeepCompareEffect(
  effect: EffectCallback,
  dependencies: DependencyList
) {
  useCustomCompareEffect(effect, dependencies, isDeepEqual);
}
