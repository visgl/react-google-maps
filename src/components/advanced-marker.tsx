/* eslint-disable complexity */
import React, {
  Children,
  CSSProperties,
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useState
} from 'react';

import {createPortal} from 'react-dom';
import {useMap} from '../hooks/use-map';
import {useMapsLibrary} from '../hooks/use-maps-library';

import type {Ref, PropsWithChildren} from 'react';
import {useMapsEventListener} from '../hooks/use-maps-event-listener';
import {usePropBinding} from '../hooks/use-prop-binding';
import {useDomEventListener} from '../hooks/use-dom-event-listener';

export interface AdvancedMarkerContextValue {
  marker: google.maps.marker.AdvancedMarkerElement;
}

export function isAdvancedMarker(
  marker: google.maps.Marker | google.maps.marker.AdvancedMarkerElement
): marker is google.maps.marker.AdvancedMarkerElement {
  return (
    (marker as google.maps.marker.AdvancedMarkerElement).content !== undefined
  );
}

function isElementNode(node: Node): node is HTMLElement {
  return node.nodeType === Node.ELEMENT_NODE;
}

/**
 * Copy of the `google.maps.CollisionBehavior` constants.
 * They have to be duplicated here since we can't wait for the maps API to load to be able to use them.
 */
export const CollisionBehavior = {
  REQUIRED: 'REQUIRED',
  REQUIRED_AND_HIDES_OPTIONAL: 'REQUIRED_AND_HIDES_OPTIONAL',
  OPTIONAL_AND_HIDES_LOWER_PRIORITY: 'OPTIONAL_AND_HIDES_LOWER_PRIORITY'
} as const;
export type CollisionBehavior =
  (typeof CollisionBehavior)[keyof typeof CollisionBehavior];

export const AdvancedMarkerContext =
  React.createContext<AdvancedMarkerContextValue | null>(null);

// [xPosition, yPosition] when the top left corner is [0, 0]
export const AdvancedMarkerAnchorPoint = {
  TOP_LEFT: ['0%', '0%'],
  TOP_CENTER: ['50%', '0%'],
  TOP: ['50%', '0%'],
  TOP_RIGHT: ['100%', '0%'],
  LEFT_CENTER: ['0%', '50%'],
  LEFT_TOP: ['0%', '0%'],
  LEFT: ['0%', '50%'],
  LEFT_BOTTOM: ['0%', '100%'],
  RIGHT_TOP: ['100%', '0%'],
  RIGHT: ['100%', '50%'],
  RIGHT_CENTER: ['100%', '50%'],
  RIGHT_BOTTOM: ['100%', '100%'],
  BOTTOM_LEFT: ['0%', '100%'],
  BOTTOM_CENTER: ['50%', '100%'],
  BOTTOM: ['50%', '100%'],
  BOTTOM_RIGHT: ['100%', '100%'],
  CENTER: ['50%', '50%']
} as const;

export type AdvancedMarkerAnchorPoint =
  (typeof AdvancedMarkerAnchorPoint)[keyof typeof AdvancedMarkerAnchorPoint];

type AdvancedMarkerEventProps = {
  onClick?: (e: google.maps.MapMouseEvent) => void;
  onMouseEnter?: (e: google.maps.MapMouseEvent['domEvent']) => void;
  onMouseLeave?: (e: google.maps.MapMouseEvent['domEvent']) => void;
  onDrag?: (e: google.maps.MapMouseEvent) => void;
  onDragStart?: (e: google.maps.MapMouseEvent) => void;
  onDragEnd?: (e: google.maps.MapMouseEvent) => void;
};

export type AdvancedMarkerProps = PropsWithChildren<
  Omit<
    google.maps.marker.AdvancedMarkerElementOptions,
    'gmpDraggable' | 'gmpClickable' | 'content' | 'map' | 'collisionBehavior'
  > &
    AdvancedMarkerEventProps & {
      draggable?: boolean;
      clickable?: boolean;
      collisionBehavior?: CollisionBehavior;
      /**
       * The anchor point for the Advanced Marker.
       * Either use one of the predefined anchor point from the "AdvancedMarkerAnchorPoint" export
       * or provide a string tuple in the form of ["xPosition", "yPosition"].
       * The position is measured from the top-left corner and
       * can be anything that can be consumed by a CSS translate() function.
       * For example in percent ("50%") or in pixels ("20px").
       */
      anchorPoint?: AdvancedMarkerAnchorPoint | [string, string];
      /**
       * A className for the content element.
       * (can only be used with HTML Marker content)
       */
      className?: string;
      /**
       * Additional styles to apply to the content element.
       */
      style?: CSSProperties;
    }
>;

type MarkerContentProps = PropsWithChildren & {
  styles?: CSSProperties;
  className?: string;
  anchorPoint?: AdvancedMarkerAnchorPoint | [string, string];
};

const MarkerContent = ({
  children,
  styles,
  className,
  anchorPoint
}: MarkerContentProps) => {
  const [xTranslation, yTranslation] =
    anchorPoint ?? AdvancedMarkerAnchorPoint['BOTTOM'];

  // The "translate(50%, 100%)" is here to counter and reset the default anchoring of the advanced marker element
  // that comes from the api
  const transformStyle = `translate(50%, 100%) translate(-${xTranslation}, -${yTranslation})`;

  return (
    // anchoring container
    <div style={{transform: transformStyle}}>
      {/* AdvancedMarker div that user can give styles and classes */}
      <div className={className} style={styles}>
        {children}
      </div>
    </div>
  );
};

export type CustomMarkerContent =
  | (HTMLDivElement & {isCustomMarker?: boolean})
  | null;

export type AdvancedMarkerRef = google.maps.marker.AdvancedMarkerElement | null;
function useAdvancedMarker(props: AdvancedMarkerProps) {
  const [marker, setMarker] =
    useState<google.maps.marker.AdvancedMarkerElement | null>(null);
  const [contentContainer, setContentContainer] =
    useState<HTMLDivElement | null>(null);

  const map = useMap();
  const markerLibrary = useMapsLibrary('marker');

  const {
    children,
    onClick,
    className,
    onMouseEnter,
    onMouseLeave,
    onDrag,
    onDragStart,
    onDragEnd,
    collisionBehavior,
    clickable,
    draggable,
    position,
    title,
    zIndex
  } = props;

  const numChildren = Children.count(children);

  // create an AdvancedMarkerElement instance and add it to the map once available
  useEffect(() => {
    if (!map || !markerLibrary) return;

    const newMarker = new markerLibrary.AdvancedMarkerElement();
    newMarker.map = map;

    setMarker(newMarker);

    // create the container for marker content if there are children
    let contentElement: CustomMarkerContent = null;
    if (numChildren > 0) {
      contentElement = document.createElement('div');

      // We need some kind of flag to identify the custom marker content
      // in the infowindow component. Choosing a custom property instead of a className
      // to not encourage users to style the marker content directly.
      contentElement.isCustomMarker = true;

      newMarker.content = contentElement;
      setContentContainer(contentElement);
    }

    return () => {
      newMarker.map = null;
      contentElement?.remove();
      setMarker(null);
      setContentContainer(null);
    };
  }, [map, markerLibrary, numChildren]);

  // When no children are present we don't have our own wrapper div
  // which usually gets the user provided className. In this case
  // we set the className directly on the marker.content element that comes
  // with the AdvancedMarker.
  useEffect(() => {
    if (!marker || !marker.content || numChildren > 0) return;

    (marker.content as HTMLElement).className = className || '';
  }, [marker, className, numChildren]);

  // copy other props
  usePropBinding(marker, 'position', position);
  usePropBinding(marker, 'title', title ?? '');
  usePropBinding(marker, 'zIndex', zIndex);
  usePropBinding(
    marker,
    'collisionBehavior',
    collisionBehavior as google.maps.CollisionBehavior
  );

  // set gmpDraggable from props (when unspecified, it's true if any drag-event
  // callbacks are specified)
  useEffect(() => {
    if (!marker) return;

    if (draggable !== undefined) marker.gmpDraggable = draggable;
    else if (onDrag || onDragStart || onDragEnd) marker.gmpDraggable = true;
    else marker.gmpDraggable = false;
  }, [marker, draggable, onDrag, onDragEnd, onDragStart]);

  // set gmpClickable from props (when unspecified, it's true if the onClick or one of
  // the hover events callbacks are specified)
  useEffect(() => {
    if (!marker) return;

    const gmpClickable =
      clickable !== undefined ||
      Boolean(onClick) ||
      Boolean(onMouseEnter) ||
      Boolean(onMouseLeave);

    // gmpClickable is only available in beta version of the
    // maps api (as of 2024-10-10)
    marker.gmpClickable = gmpClickable;

    // enable pointer events for the markers with custom content
    if (gmpClickable && marker?.content && isElementNode(marker.content)) {
      marker.content.style.pointerEvents = 'none';

      if (marker.content.firstElementChild) {
        (marker.content.firstElementChild as HTMLElement).style.pointerEvents =
          'all';
      }
    }
  }, [marker, clickable, onClick, onMouseEnter, onMouseLeave]);

  useMapsEventListener(marker, 'click', onClick);
  useMapsEventListener(marker, 'drag', onDrag);
  useMapsEventListener(marker, 'dragstart', onDragStart);
  useMapsEventListener(marker, 'dragend', onDragEnd);

  useDomEventListener(marker?.element, 'mouseenter', onMouseEnter);
  useDomEventListener(marker?.element, 'mouseleave', onMouseLeave);

  return [marker, contentContainer] as const;
}

export const AdvancedMarker = forwardRef(
  (props: AdvancedMarkerProps, ref: Ref<AdvancedMarkerRef>) => {
    const {children, style, className, anchorPoint} = props;
    const [marker, contentContainer] = useAdvancedMarker(props);

    const advancedMarkerContextValue: AdvancedMarkerContextValue | null =
      useMemo(() => (marker ? {marker} : null), [marker]);

    useImperativeHandle(ref, () => marker, [marker]);

    if (!contentContainer) return null;

    return (
      <AdvancedMarkerContext.Provider value={advancedMarkerContextValue}>
        {createPortal(
          <MarkerContent
            anchorPoint={anchorPoint}
            styles={style}
            className={className}>
            {children}
          </MarkerContent>,
          contentContainer
        )}
      </AdvancedMarkerContext.Provider>
    );
  }
);

export function useAdvancedMarkerRef() {
  const [marker, setMarker] =
    useState<google.maps.marker.AdvancedMarkerElement | null>(null);

  const refCallback = useCallback((m: AdvancedMarkerRef | null) => {
    setMarker(m);
  }, []);

  return [refCallback, marker] as const;
}
