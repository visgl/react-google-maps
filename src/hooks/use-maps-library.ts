import {useContext, useEffect} from 'react';

import {APIProviderContext} from '../components/api-provider';
import {useApiIsLoaded} from './use-api-is-loaded';

interface ApiLibraries {
  core: google.maps.CoreLibrary;
  maps: google.maps.MapsLibrary;
  places: google.maps.PlacesLibrary;
  geocoding: google.maps.GeocodingLibrary;
  routes: google.maps.RoutesLibrary;
  marker: google.maps.MarkerLibrary;
  geometry: google.maps.GeometryLibrary;
  elevation: google.maps.ElevationLibrary;
  streetView: google.maps.StreetViewLibrary;
  journeySharing: google.maps.JourneySharingLibrary;
  drawing: google.maps.DrawingLibrary;
  visualization: google.maps.VisualizationLibrary;
  maps3d: google.maps.Maps3DLibrary;
}

export function useMapsLibrary<
  K extends keyof ApiLibraries,
  V extends ApiLibraries[K]
>(name: K): V | null;

export function useMapsLibrary(name: string) {
  const apiIsLoaded = useApiIsLoaded();
  const ctx = useContext(APIProviderContext);

  useEffect(() => {
    if (!apiIsLoaded || !ctx) return;

    // Trigger loading the libraries via our proxy-method.
    // The returned promise is ignored, since importLibrary will update loadedLibraries
    // list in the context, triggering a re-render.
    void ctx.importLibrary(name);
  }, [apiIsLoaded, ctx, name]);

  return ctx?.loadedLibraries[name] || null;
}
