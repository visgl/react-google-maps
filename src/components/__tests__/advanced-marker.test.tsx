import React from 'react';

import {initialize, mockInstances} from '@googlemaps/jest-mocks';
import {cleanup, queryByTestId, render} from '@testing-library/react';
import '@testing-library/jest-dom';

import {AdvancedMarker, AdvancedMarkerAnchorPoint} from '../advanced-marker';
import {useMap} from '../../hooks/use-map';
import {useMapsLibrary} from '../../hooks/use-maps-library';

import {waitForMockInstance} from './__utils__/wait-for-mock-instance';

jest.mock('../../hooks/use-map');
jest.mock('../../hooks/use-maps-library');

let createMarkerSpy: jest.Mock;
let useMapMock: jest.MockedFn<typeof useMap>;
let useMapsLibraryMock: jest.MockedFn<typeof useMapsLibrary>;

let markerLib: google.maps.MarkerLibrary;
let mapInstance: google.maps.Map;

beforeEach(async () => {
  initialize();
  jest.clearAllMocks();

  // mocked versions of the useMap and useMapsLibrary functions
  useMapMock = jest.mocked(useMap);
  useMapsLibraryMock = jest.mocked(useMapsLibrary);

  // load the marker-lib that can be returned from the useMapsLibrary mock
  markerLib = (await google.maps.importLibrary(
    'marker'
  )) as google.maps.MarkerLibrary;

  // custom implementation of the AdvancedMarkerElement that has a properly
  // initialized content and an observeable constructor.
  createMarkerSpy = jest.fn();
  const AdvancedMarkerElement = class extends google.maps.marker
    .AdvancedMarkerElement {
    constructor(o?: google.maps.marker.AdvancedMarkerElementOptions) {
      createMarkerSpy(o);
      super(o);

      // @googlemaps/js-jest-mocks doesn't initialize the .content property
      // as the real implementation does (this would normally be a pin-element,
      // but for our purposes a div should suffice)
      this.content = document.createElement('div');
    }
  };

  // the element has to be registered for this override to work, but since we
  // can neither unregister nor re-register an element, a randomized name is
  // used for the element.
  customElements.define(
    `gmp-advanced-marker-${Math.random().toString(36).slice(2)}`,
    AdvancedMarkerElement
  );

  google.maps.marker.AdvancedMarkerElement = markerLib.AdvancedMarkerElement =
    AdvancedMarkerElement;
});

afterEach(() => {
  cleanup();
});

test('creates marker instance once map is ready', async () => {
  useMapsLibraryMock.mockReturnValue(null);
  useMapMock.mockReturnValue(null);

  const {rerender} = render(<AdvancedMarker position={{lat: 0, lng: 0}} />);

  expect(useMapsLibraryMock).toHaveBeenCalledWith('marker');
  expect(createMarkerSpy).not.toHaveBeenCalled();

  useMapsLibraryMock.mockImplementation(name => {
    expect(name).toEqual('marker');
    return markerLib;
  });
  const mockMap = new google.maps.Map(document.createElement('div'));
  useMapMock.mockReturnValue(mockMap);

  rerender(<AdvancedMarker position={{lat: 1, lng: 2}} />);

  expect(createMarkerSpy).toHaveBeenCalled();
});

describe('map and marker-library loaded', () => {
  beforeEach(() => {
    useMapsLibraryMock.mockImplementation(name => {
      expect(name).toEqual('marker');
      return markerLib;
    });

    mapInstance = new google.maps.Map(document.createElement('div'));
    useMapMock.mockReturnValue(mapInstance);
  });

  test('marker should be initialized', async () => {
    render(<AdvancedMarker position={{lat: 0, lng: 0}} />);

    expect(createMarkerSpy).toHaveBeenCalledTimes(1);
    expect(
      mockInstances.get(google.maps.marker.AdvancedMarkerElement)
    ).toHaveLength(1);
  });

  test('marker should have and update its position', async () => {
    const {rerender} = render(<AdvancedMarker position={{lat: 1, lng: 0}} />);
    const marker = await waitForMockInstance(
      google.maps.marker.AdvancedMarkerElement
    );

    expect(marker.position).toEqual({lat: 1, lng: 0});

    rerender(<AdvancedMarker position={{lat: 2, lng: 2}} />);

    expect(marker.position).toEqual({lat: 2, lng: 2});
  });

  test('marker class is set correctly without html content', async () => {
    render(
      <AdvancedMarker
        position={{lat: 1, lng: 2}}
        className={'classname-test'}
      />
    );

    const marker = mockInstances
      .get(google.maps.marker.AdvancedMarkerElement)
      .at(0) as google.maps.marker.AdvancedMarkerElement;

    expect(marker.content).toHaveClass('classname-test');
  });

  test('marker class and style are set correctly with html content', async () => {
    render(
      <AdvancedMarker
        position={{lat: 1, lng: 2}}
        className={'classname-test'}
        style={{width: 200}}>
        <div data-testid={'marker-content'}>Marker Content!</div>
      </AdvancedMarker>
    );

    const marker = mockInstances
      .get(google.maps.marker.AdvancedMarkerElement)
      .at(0) as google.maps.marker.AdvancedMarkerElement;

    const advancedMarkerWithClass = (
      marker.content as HTMLElement
    ).querySelector('.classname-test');

    expect(advancedMarkerWithClass).toBeTruthy();
    expect(advancedMarkerWithClass).toHaveStyle('width: 200px');

    expect(
      queryByTestId(marker.content as HTMLElement, 'marker-content')
    ).toBeTruthy();
  });

  test.todo('marker should work with options');
  test.todo('marker should have a click listener');

  describe('anchoring with modern API', () => {
    beforeEach(() => {
      google.maps.version = '3.62.9';
    });

    test('anchorLeft/anchorTop should have precedence over anchorPoint', async () => {
      const consoleWarnSpy = jest
        .spyOn(console, 'warn')
        .mockImplementation(() => {});

      render(
        <AdvancedMarker
          position={{lat: 0, lng: 0}}
          anchorLeft={'10px'}
          anchorTop={'20px'}
          anchorPoint={['10%', '20%']}>
          <div />
        </AdvancedMarker>
      );
      const marker = await waitForMockInstance(
        google.maps.marker.AdvancedMarkerElement
      );

      expect(marker.anchorLeft).toBe('10px');
      expect(marker.anchorTop).toBe('20px');
      expect(consoleWarnSpy.mock.calls).toMatchSnapshot();

      consoleWarnSpy.mockRestore();
    });

    test('anchorPoint should be used as fallback', async () => {
      render(
        <AdvancedMarker
          position={{lat: 0, lng: 0}}
          anchorPoint={['12%', '34%']}>
          <div />
        </AdvancedMarker>
      );
      const marker = await waitForMockInstance(
        google.maps.marker.AdvancedMarkerElement
      );

      expect(marker.anchorLeft).toBe('calc(-1 * 12%)');
      expect(marker.anchorTop).toBe('calc(-1 * 34%)');
    });
  });

  describe('anchoring with legacy API', () => {
    beforeEach(() => {
      google.maps.version = '3.61.0';
    });

    test('anchorPoint is applied as css transform', async () => {
      render(
        <AdvancedMarker
          position={{lat: 0, lng: 0}}
          anchorPoint={AdvancedMarkerAnchorPoint.CENTER}>
          <div />
        </AdvancedMarker>
      );
      const marker = await waitForMockInstance(
        google.maps.marker.AdvancedMarkerElement
      );

      expect(marker.content).toBeInstanceOf(HTMLElement);
      expect((marker.content as HTMLElement).style.transform).toBe(
        'translate(50%, 100%) translate(calc(-1 * 50%), calc(-1 * 50%))'
      );
    });

    test('should warn when using anchorLeft/Top', async () => {
      const consoleWarnSpy = jest
        .spyOn(console, 'warn')
        .mockImplementation(() => {});

      render(
        <AdvancedMarker
          position={{lat: 0, lng: 0}}
          anchorLeft={'10px'}
          anchorTop={'20px'}>
          <div />
        </AdvancedMarker>
      );
      await waitForMockInstance(google.maps.marker.AdvancedMarkerElement);

      expect(consoleWarnSpy.mock.calls).toMatchSnapshot();

      consoleWarnSpy.mockRestore();
    });
  });
});
