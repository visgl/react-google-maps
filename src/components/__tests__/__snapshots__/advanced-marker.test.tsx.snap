// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`map and marker-library loaded anchoring with legacy API should warn when using anchorLeft/Top 1`] = `
[
  [
    "AdvancedMarker: The anchorLeft and anchorTop props are only supported in Google Maps API version 3.62 and above. The current version is 3.61.0.",
  ],
]
`;

exports[`map and marker-library loaded anchoring with modern API anchorLeft/anchorTop should have precedence over anchorPoint 1`] = `
[
  [
    "AdvancedMarker: the anchorPoint prop is ignored when anchorLeft and/or anchorTop are set.",
  ],
]
`;
